include/albatross/src/covariance_functions/traits.hpp

```diff
@@ -32,14 +32,41 @@ HAS_METHOD(_call_impl);
 template <typename U, typename... Args>
 class has_possible_call_impl : public has__call_impl<U, Args &...> {};
 
+/*
+ * has_valid_cov_caller
+ */
 HAS_METHOD_WITH_RETURN_TYPE(call);
 
-template <typename U, typename Caller, typename... Args>
+template <typename CovFunc, typename Caller, typename... Args>
 class has_valid_cov_caller
     : public has_call_with_return_type<Caller, double,
-                                       typename const_ref<U>::type,
+                                       typename const_ref<CovFunc>::type,
                                        typename const_ref<Args>::type...> {};
 
+/*
+ * has_valid_cross_cov_caller
+ */
+template <typename U, typename Caller, typename A, typename B>
+class has_valid_cross_cov_caller {
+public:
+  static constexpr bool value = has_valid_cov_caller<U, Caller, A, A>::value &&
+                                has_valid_cov_caller<U, Caller, A, B>::value &&
+                                has_valid_cov_caller<U, Caller, B, B>::value;
+};
+
+template <typename U, typename Caller, typename A, typename B>
+struct has_valid_cross_cov_caller<U, Caller, A, variant<B>>
+    : public has_valid_cross_cov_caller<U, Caller, A, B> {
+  ;
+};
+
+template <typename U, typename Caller, typename A, typename B, typename... Ts>
+struct has_valid_cross_cov_caller<U, Caller, A, variant<B, Ts...>> {
```

Comment:
You'd know better than me, but it seems like it doesn't make sense to say if there's valid cross cov caller if not all of the variant types have a caller function.

Reply:
Yeah, great catch.  This forced me to revise some of the tests as well.  Thanks.